### PR TITLE
feat(dtfs2-6966): add audit logs to update facility endpoint

### PR DESCRIPTION
**Pull request by AlexBramhill**
## Introduction :pencil2:
We need to add audit details to portal endpoints

## Resolution :heavy_check_mark:
This PR adds audit details to the `update facility` endpoint

### CODE REVIEW

File: dtfs-central-api/api-tests/v1/facilities/facility-put.api-test.js [hunk 1]
```diff
@@ -40,93 +44,93 @@ describe('/v1/portal/facilities', () => {
...
-        user: MOCK_PORTAL_USER,
-      };
-
-      const { body, status } = await api.put(updatedFacility).to(`/v1/portal/facilities/${createdFacility._id}`);
+      const { body, status } = await api.put(aValidUpdateRequest).to(`/v1/portal/facilities/${createdFacility._id}`);
 
       expect(status).toEqual(200);
       expect(typeof body.updatedAt).toEqual('number');
-      expect(body.value).toEqual(updatedFacility.value);
+      expect(body.value).toEqual(aValidUpdateRequest.facilityUpdate.value);
```
Comment (MarRobSoftwire) [inline dtfs-central-api/api-tests/v1/facilities/facility-put.api-test.js:92]: Odd that this doesn't include the audit record

File: dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js [hunk 2]
```diff
@@ -12,51 +13,56 @@ const withoutId = (obj) => {
   return cleanedObject;
 };
 
-const updateFacility = async (facilityId, facilityBody, dealId, user, routePath) => {
-  if (ObjectId.isValid(dealId) && ObjectId.isValid(facilityId)) {
-    const collection = await db.getCollection(MONGO_DB_COLLECTIONS.FACILITIES);
+const updateFacility = async (facilityId, facilityUpdate, dealId, user, routePath, auditDetails) => {
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js:16]: Would prefer this to destructure params from an object

```suggestion
const updateFacility = async {(facilityId, facilityUpdate, dealId, user, routePath, auditDetails}) => {
```

Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js:16]: Would prefer this to destructure params from an object

```suggestion
const updateFacility = async {(facilityId, facilityUpdate, dealId, user, routePath, auditDetails}) => {
```

File: dtfs-central-api/src/v1/routes/portal-routes.js [hunk 3]
```diff
@@ -408,7 +408,13 @@ portalRouter.route('/facilities/:id').get(getFacilityController.findOneFacilityG
...
  *           schema:
  *             type: object
- *             example: { aNewField: true }
+ *             properties:
+ *               facilityUpdate:
+ *                 type: object
+ *               user:
+ *                 type: object
+ *               auditDetails:
+ *                 $ref: '#/definitions/portalAuditDetails'
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/routes/portal-routes.js:417]:
```suggestion
 *               auditDetails:
 *                 type: object
 *                 $ref: '#/definitions/portalAuditDetails'
```

File: dtfs-central-api/src/v1/routes/portal-routes.js [hunk 4]
```diff
@@ -770,7 +776,6 @@ portalRouter.route('/gef/facilities').get(getGefFacilitiesController.findAllFaci
  *         application/json:
  *           schema:
  *             type: object
- *             example: { aNewField: true }
```
Comment (MarRobSoftwire) [inline dtfs-central-api/src/v1/routes/portal-routes.js:773]: How come this has been deleted?

File: trade-finance-manager-api/src/v1/api.js [hunk 5]
```diff
@@ -150,7 +150,7 @@ const updatePortalFacility = async (facilityId, update) => {
       method: 'put',
       url: `${DTFS_CENTRAL_API_URL}/v1/portal/facilities/${facilityId}`,
       headers: headers.central,
-      data: update,
+      data: { facilityUpdate: update },
```
Comment (MarRobSoftwire) [inline trade-finance-manager-api/src/v1/api.js:153]: Does this not need auditDetails on it?

File: utils/mock-data-loader/centralApi.js [hunk 6]
```diff
@@ -49,7 +49,7 @@ const updateFacility = async (facilityId, facilityUpdate, user) => {
     url: `${DTFS_CENTRAL_API_URL}/v1/portal/facilities/${facilityId}`,
     headers,
     data: {
-      ...facilityUpdate,
+      facilityUpdate,
```
Comment (MarRobSoftwire) [inline utils/mock-data-loader/centralApi.js:52]: Do we not need audit details on this? Suprised the validation isn't throwing an error

File: dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js [hunk 7]
```diff
@@ -12,51 +13,56 @@ const withoutId = (obj) => {
   return cleanedObject;
 };
 
-const updateFacility = async (facilityId, facilityBody, dealId, user, routePath) => {
-  if (ObjectId.isValid(dealId) && ObjectId.isValid(facilityId)) {
-    const collection = await db.getCollection(MONGO_DB_COLLECTIONS.FACILITIES);
+const updateFacility = async ({ facilityId, facilityUpdate, dealId, user, routePath, auditDetails }) => {
+  if (!ObjectId.isValid(dealId) || !ObjectId.isValid(facilityId)) {
+    return { status: 400, message: 'Invalid Deal or Facility Id' };
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js:18]: yeaaahhh early return

File: dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js [hunk 8]
```diff
@@ -12,51 +13,56 @@ const withoutId = (obj) => {
...
-      await updateDealEditedByPortal(dealId, user);
-    }
+  const { value: updatedFacility } = findAndUpdateResponse;
+
+  if (routePath === PORTAL_ROUTE && user) {
+    // update the deal so that the user that has edited this facility,
+    // is also marked as editing the associated deal
 
-    return updatedFacility;
+    await updateDealEditedByPortal(dealId, user);
```
Comment (ChristianMcc) [inline dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js:38]: I haven't traced this but does this cause an update to the deal or facility as well which needs an audit record?

**Review by ChristianMcc — COMMENTED** [review]
Nice, minor comment baout the editedBy part

File: dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js (hunk 8 above)
Comment (AlexBramhill) [reply dtfs-central-api/src/v1/controllers/portal/facility/update-facility.controller.js:38]: Spoilers! Theres a PR for this here once this one is merged https://github.com/UK-Export-Finance/dtfs2/pull/3252/files (note its currently sat on this one -- so its larger than it should be)

**Review by ChristianMcc — APPROVED** [review]

**Pull request merged**